### PR TITLE
Use ahrs Madgwick for orientation

### PR DESCRIPTION
**Pull request by schmidma**
## Why? What?

Use a Madgwick filter to fuse gyroscope and accelerometer.

## ToDo / Known Issues

none

## Ideas for Next Iterations (Not This PR)

none

## How to Test

observe the horizon or the euler angles (#1106) with twix

### CODE REVIEW

File: crates/control/src/orientation_filter.rs [hunk 1]
```diff
@@ -24,16 +33,16 @@ pub struct CreationContext {}
...
 pub struct CycleContext {
     sensor_data: Input<SensorData, "sensor_data">,
-    cycle_time: Input<CycleTime, "cycle_time">,
-    sole_pressure: Input<SolePressure, "sole_pressure">,
-
-    orientation_filter_parameters: Parameter<Parameters, "orientation_filter">,
+    filter_gain: Parameter<f32, "orientation_filter.filter_gain">,
+    calibration_steady_threshold: Parameter<f32, "orientation_filter.calibration_steady_threshold">,
+    calibration_smoothing_factor: Parameter<f32, "orientation_filter.calibration_smoothing_factor">,
+    num_calibration_cycles: Parameter<usize, "orientation_filter.num_calibration_cycles">,
```
Comment (oleflb) [inline crates/control/src/orientation_filter.rs:39]: please change `num_` to `number_of_` or similar

File: crates/filtering/src/madgwick.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+use num_traits::{One, Zero};
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// https://github.com/jmagnuson/ahrs-rs
+pub trait Madgwick<T> {
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:13]: I don't know about having `Madgwick` generic. In my opinion it's enough to implement for `UnitQuaternion<f32>` to remove all the generics. In the rest of our codebase we use `f32` anyway

File: crates/filtering/src/madgwick.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:19]: That would also allow to have this properly typed as `Duration`

File: crates/filtering/src/madgwick.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// https://github.com/jmagnuson/ahrs-rs
+pub trait Madgwick<T> {
+    fn update_with_imu(
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:14]: please don't abbreviate

File: crates/filtering/src/madgwick.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero>;
+
+    fn update_with_gyro(&mut self, gyroscope: &Vector3<T>, sample_period: T);
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:22]: please don't abbreviate

File: crates/filtering/src/madgwick.rs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+impl<T> Madgwick<T> for UnitQuaternion<T>
+where
+    T: Scalar + SimdValue + RealField + One + Zero + Copy,
+{
+    #[allow(non_snake_case)]
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:29]: isn't his snakecase anyway?

File: crates/filtering/src/madgwick.rs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero> {
+        let q = self.as_ref();
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:37]: I know this whole function is probably just copied, but please think about proper names for all the variables used here.

File: crates/filtering/src/madgwick.rs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        let half: T = nalgebra::convert(0.5);
+
+        // Normalize accelerometer measurement
+        let Some(accel) = accelerometer.try_normalize(zero) else {
+            return Err(AccelerometerNormZero);
+        };
+
+        // Gradient descent algorithm corrective step
+        #[rustfmt::skip]
+        let F = Vector4::new(
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:51]: please use vector! macro instead

File: crates/filtering/src/madgwick.rs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            two*(       q[0]*q[2] - q[3]*q[1]) - accel[0],
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:52]: please prefer `i,j,k,w` indexing to `[0], [1], [2], [3]`

File: crates/filtering/src/madgwick.rs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            two*(       q[3]*q[0] + q[1]*q[2]) - accel[1],
+            two*(half - q[0]*q[0] - q[1]*q[1]) - accel[2],
+            zero
+        );
+
+        #[rustfmt::skip]
+        let J_t = Matrix4::new(
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:59]: what do you think about using `matrix!` here?

File: crates/filtering/src/madgwick.rs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+        // Try to normalize step, falling back to gyro update if not possible
+        let Some(step) = (J_t * F).try_normalize(zero) else {
+            self.update_with_gyro(gyroscope, sample_period);
+            return Ok(());
+        };
+
+        // Compute rate of change of quaternion
+        let q_dot = (q * Quaternion::from_parts(zero, *gyroscope)) * half
+            - Quaternion::new(step[0], step[1], step[2], step[3]) * filter_gain;
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:74]: `Quaternion::from_vector`

**Review by oleflb — COMMENTED** [review]
Not yet tested, only code review

File: crates/filtering/src/madgwick.rs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+};
+use num_traits::{One, Zero};
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// https://github.com/jmagnuson/ahrs-rs
```
Comment (knoellle) [inline crates/filtering/src/madgwick.rs:12]: See #1110 

```suggestion
/// <https://github.com/jmagnuson/ahrs-rs>
```

**Comment by h3ndrk** [comment]
Sorry for being pedantic but this PRs description does not describe the Why

File: crates/control/src/orientation_filter.rs (hunk 1 above)
Comment (schmidma) [reply crates/control/src/orientation_filter.rs:39]: ok

File: crates/filtering/src/madgwick.rs (hunk 2 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:13]: it is generic already, and it was from the source I copied it from. Is there any reason to drop this existing implementation?

File: crates/filtering/src/madgwick.rs (hunk 4 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:14]: are you sure?

File: crates/filtering/src/madgwick.rs (hunk 5 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:22]: done

File: crates/filtering/src/madgwick.rs (hunk 6 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:29]: No, there are several variables that are not snake case

File: crates/filtering/src/madgwick.rs (hunk 7 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:37]: I actually did not have a look at the algorithm at all... That would be a lot of effort to first understand how it works, to then subsequently name all variables appropriately

File: crates/filtering/src/madgwick.rs (hunk 8 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:51]: done

File: crates/filtering/src/madgwick.rs (hunk 10 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:59]: done

File: crates/filtering/src/madgwick.rs (hunk 9 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:52]: done

File: crates/filtering/src/madgwick.rs (hunk 11 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:74]: done

File: crates/filtering/src/madgwick.rs (hunk 3 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs:19]: yes

File: crates/filtering/src/madgwick.rs (hunk 12 above)
Comment (schmidma) [reply crates/filtering/src/madgwick.rs]: done

File: crates/filtering/src/madgwick.rs (hunk 6 above)
Comment (oleflb) [reply crates/filtering/src/madgwick.rs:25]: ah I missed this applies to whole function body

**Review by oleflb — APPROVED** [review]

**Review by oleflb — CHANGES_REQUESTED** [review]
Need to block, just noticed something

File: crates/filtering/src/madgwick.rs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        };
+
+        // Gradient descent algorithm corrective step
+        #[rustfmt::skip]
+        let F = vector![
+            2.0 * (      q.i * q.k - q.w * q.j) - accel.x,
+            2.0 * (      q.w * q.i + q.j * q.k) - accel.y,
+            2.0 * (0.5 - q.i * q.i - q.j * q.j) - accel.z,
+            0.0
+        ];
```
Comment (oleflb) [inline crates/filtering/src/madgwick.rs:47]:
```suggestion
let reference = Quaternion::from_vector(Vector3::z());
let F = (q.conjugate() * reference * q).as_vector()  - accel.to_homogeneous()
```

Comment (oleflb) [reply crates/filtering/src/madgwick.rs:47]: talked to @schmidma. Although this suggestion does not change the filter, we will implement these steps later, tracked by #1133

**Review by oleflb — APPROVED** [review]

**Pull request merged**